File: ios/TougeDash/History/IncidentCaptureEngine.swift
import Foundation
import SwiftData

struct IncidentCaptureEngine: Sendable {
    struct Configuration: Equatable, Sendable {
        var sampleInterval: TimeInterval = 1.0 / 25.0
        var preTriggerDuration: TimeInterval = 30
        var postTriggerDuration: TimeInterval = 60
        var cooldown: TimeInterval = 300
        var lowOilPressureEnabled = true
        var lowOilPressureBar = 1.5
        var lowOilMinimumRPM = 3_000.0
        var lowOilDuration: TimeInterval = 0.75
        var leanUnderBoostEnabled = true
        var leanAFR = 13.5
        var leanMinimumBoostBar = 0.5
        var leanDuration: TimeInterval = 0.5
        var overboostEnabled = true
        var overboostBar = 1.5
        var overboostDuration: TimeInterval = 0.5
        var highCoolantTemperatureEnabled = true
        var coolantTemperatureCelsius = 110.0
        var coolantDuration: TimeInterval = 2
        var highOilTemperatureEnabled = true
        var oilTemperatureCelsius = 120.0
        var oilTemperatureDuration: TimeInterval = 2
        var lowBatteryVoltageEnabled = true
        var lowBatteryVoltage = 11.5
        var lowBatteryMinimumRPM = 800.0
        var lowBatteryDuration: TimeInterval = 3
        var lowFuelPressureEnabled = false
        var lowFuelPressureBar = 2.5
        var lowFuelPressureMinimumRPM = 1_500.0
        var lowFuelPressureDuration: TimeInterval = 1

        static let standard = Configuration()
    }

    struct CompletedIncident: Sendable {
        let id: UUID
        let kind: IncidentKind
        let severity: IncidentSeverity
        let triggeredAt: Date
        let thresholdValue: Double
        let triggerValue: Double
        let triggerUnit: String
        let conditionDuration: TimeInterval
        let samples: [CapturedTelemetryPoint]
    }

    private struct RuleMatch: Sendable {
        let kind: IncidentKind
        let severity: IncidentSeverity
        let value: Double
        let threshold: Double
        let unit: String
        let debounce: TimeInterval
    }

    private struct ActiveCapture: Sendable {
        let id: UUID
        let match: RuleMatch
        let triggeredAt: Date
        let finishAt: Date
        let conditionStartedAt: Date
        var conditionEndedAt: Date?
        var samples: [CapturedTelemetryPoint]
    }

    private(set) var configuration: Configuration
    private var preTriggerBuffer: [CapturedTelemetryPoint] = []
    private var activeCaptures: [IncidentKind: ActiveCapture] = [:]
    private var conditionStartedAt: [IncidentKind: Date] = [:]
    private var lastTriggeredAt: [IncidentKind: Date] = [:]
    private var lastAcceptedAt = Date.distantPast

    init(configuration: Configuration = .standard) {
        self.configuration = configuration
    }

    mutating func ingest(_ point: CapturedTelemetryPoint) -> [CompletedIncident] {
        guard point.timestamp.timeIntervalSince(lastAcceptedAt) >= configuration.sampleInterval else { return [] }
        lastAcceptedAt = point.timestamp

        for kind in activeCaptures.keys {
            activeCaptures[kind]?.samples.append(point)
        }

        var completed: [CompletedIncident] = []
        for (kind, capture) in activeCaptures where point.timestamp >= capture.finishAt {
            completed.append(complete(capture))
            activeCaptures.removeValue(forKey: kind)
        }

        preTriggerBuffer.append(point)
        let oldestAllowed = point.timestamp.addingTimeInterval(-configuration.preTriggerDuration)
        if let firstValid = preTriggerBuffer.firstIndex(where: { $0.timestamp >= oldestAllowed }), firstValid > 0 {
            preTriggerBuffer.removeFirst(firstValid)
        }

        let matches = matchingRules(for: point)
        let matchingKinds = Set(matches.map(\.kind))
        for kind in IncidentKind.allCases where !matchingKinds.contains(kind) {
            if activeCaptures[kind]?.conditionEndedAt == nil {
                activeCaptures[kind]?.conditionEndedAt = point.timestamp
            }
            conditionStartedAt.removeValue(forKey: kind)
        }
        for match in matches {
            if conditionStartedAt[match.kind] == nil {
                conditionStartedAt[match.kind] = point.timestamp
            }
            guard activeCaptures[match.kind] == nil,
                  let startedAt = conditionStartedAt[match.kind],
                  point.timestamp.timeIntervalSince(startedAt) >= match.debounce,
                  lastTriggeredAt[match.kind].map({ point.timestamp.timeIntervalSince($0) >= configuration.cooldown }) ?? true else {
                continue
            }
            activeCaptures[match.kind] = ActiveCapture(
                id: UUID(),
                match: match,
                triggeredAt: point.timestamp,
                finishAt: point.timestamp.addingTimeInterval(configuration.postTriggerDuration),
                conditionStartedAt: startedAt,
                conditionEndedAt: nil,
                samples: preTriggerBuffer
            )
            lastTriggeredAt[match.kind] = point.timestamp
        }
        return completed.sorted { $0.triggeredAt < $1.triggeredAt }
    }

    mutating func finishActiveCaptures() -> [CompletedIncident] {
        let completed = activeCaptures.values.map(complete).sorted { $0.triggeredAt < $1.triggeredAt }
        activeCaptures.removeAll()
        conditionStartedAt.removeAll()
        return completed
    }

    mutating func reset() {
        preTriggerBuffer.removeAll(keepingCapacity: true)
        activeCaptures.removeAll()
        conditionStartedAt.removeAll()
        lastTriggeredAt.removeAll()
        lastAcceptedAt = .distantPast
    }

    mutating func updateConfiguration(_ configuration: Configuration) {
        guard self.configuration != configuration else { return }
        self.configuration = configuration
        conditionStartedAt.removeAll()
    }

    private func complete(_ capture: ActiveCapture) -> CompletedIncident {
        CompletedIncident(
            id: capture.id,
            kind: capture.match.kind,
            severity: capture.match.severity,
            triggeredAt: capture.triggeredAt,
            thresholdValue: capture.match.threshold,
            triggerValue: capture.match.value,
            triggerUnit: capture.match.unit,
            conditionDuration: max(
                0,
                (capture.conditionEndedAt ?? capture.samples.last?.timestamp ?? capture.triggeredAt)
                    .timeIntervalSince(capture.conditionStartedAt)
            ),
            samples: capture.samples
        )
    }

    private func matchingRules(for point: CapturedTelemetryPoint) -> [RuleMatch] {
        var matches: [RuleMatch] = []
        if configuration.lowOilPressureEnabled,
           point.rpm >= configuration.lowOilMinimumRPM,
           point.oilPressureBar > 0,
           point.oilPressureBar < configuration.lowOilPressureBar {
            matches.append(RuleMatch(
                kind: .lowOilPressure,
                severity: .critical,
                value: point.oilPressureBar,
                threshold: configuration.lowOilPressureBar,
                unit: "bar",
                debounce: configuration.lowOilDuration
            ))
        }
        if configuration.leanUnderBoostEnabled,
           point.boostBar >= configuration.leanMinimumBoostBar,
           point.afr > configuration.leanAFR {
            matches.append(RuleMatch(
                kind: .leanUnderBoost,
                severity: .critical,
                value: point.afr,
                threshold: configuration.leanAFR,
                unit: "AFR",
                debounce: configuration.leanDuration
            ))
        }
        if configuration.overboostEnabled, point.boostBar > configuration.overboostBar {
            matches.append(RuleMatch(
                kind: .overboost,
                severity: .critical,
                value: point.boostBar,
                threshold: configuration.overboostBar,
                unit: "bar",
                debounce: configuration.overboostDuration
            ))
        }
        if configuration.highCoolantTemperatureEnabled,
           point.coolantCelsius >= configuration.coolantTemperatureCelsius {
            matches.append(RuleMatch(
                kind: .highCoolantTemperature,
                severity: .critical,
                value: point.coolantCelsius,
                threshold: configuration.coolantTemperatureCelsius,
                unit: "°C",
                debounce: configuration.coolantDuration
            ))
        }
        if configuration.highOilTemperatureEnabled,
           point.oilTemperatureCelsius >= configuration.oilTemperatureCelsius {
            matches.append(RuleMatch(
                kind: .highOilTemperature,
                severity: .critical,
                value: point.oilTemperatureCelsius,
                threshold: configuration.oilTemperatureCelsius,
                unit: "°C",
                debounce: configuration.oilTemperatureDuration
            ))
        }
        if configuration.lowBatteryVoltageEnabled,
           point.rpm >= configuration.lowBatteryMinimumRPM,
           point.batteryVoltage > 0,
           point.batteryVoltage < configuration.lowBatteryVoltage {
            matches.append(RuleMatch(
                kind: .lowBatteryVoltage,
                severity: .warning,
                value: point.batteryVoltage,
                threshold: configuration.lowBatteryVoltage,
                unit: "V",
                debounce: configuration.lowBatteryDuration
            ))
        }
        if configuration.lowFuelPressureEnabled,
           point.rpm >= configuration.lowFuelPressureMinimumRPM,
           point.fuelPressureBar > 0,
           point.fuelPressureBar < configuration.lowFuelPressureBar {
            matches.append(RuleMatch(
                kind: .lowFuelPressure,
                severity: .critical,
                value: point.fuelPressureBar,
                threshold: configuration.lowFuelPressureBar,
                unit: "bar",
                debounce: configuration.lowFuelPressureDuration
            ))
        }
        return matches
    }
}

@MainActor
final class TelemetryIncidentRecorder: ObservableObject {
    @Published private(set) var pendingIncidentCount = 0
    var onIncidentStored: ((Int) -> Void)?

    private let context: ModelContext
    private let locationTracker: LocationTrackingService
    private var vehicleID: UUID
    private var engine = IncidentCaptureEngine()
    private let alertRules: VehicleAlertRuleStore
    private var lastSessionID: UUID?

    init(
        container: ModelContainer,
        locationTracker: LocationTrackingService,
        alertRules: VehicleAlertRuleStore
    ) {
        context = ModelContext(container)
        context.autosaveEnabled = false
        self.locationTracker = locationTracker
        self.alertRules = alertRules
        vehicleID = LocalVehicleIdentity.resolve()
        alertRules.activateVehicle(vehicleID)
        refreshPendingCount()
    }

    func activateVehicle(_ id: UUID) {
        guard id != vehicleID else { return }
        persist(engine.finishActiveCaptures(), sessionID: lastSessionID)
        vehicleID = id
        alertRules.activateVehicle(id)
        lastSessionID = nil
        engine.reset()
        refreshPendingCount()
    }

    func record(_ snapshot: TelemetrySnapshot, sessionID: UUID, at timestamp: Date = .now) {
        lastSessionID = sessionID
        engine.updateConfiguration(alertRules.rules(for: vehicleID).incidentConfiguration)
        let location = freshLocation(at: timestamp)
        let point = CapturedTelemetryPoint(snapshot: snapshot, timestamp: timestamp, location: location)
        persist(engine.ingest(point), sessionID: sessionID)
    }

    func finish(sessionID: UUID?) {
        persist(engine.finishActiveCaptures(), sessionID: sessionID ?? lastSessionID)
        try? context.save()
    }

    private func persist(_ completed: [IncidentCaptureEngine.CompletedIncident], sessionID: UUID?) {
        guard let sessionID else { return }
        for capture in completed {
            context.insert(DriveIncident(
                id: capture.id,
                vehicleID: vehicleID,
                sessionID: sessionID,
                kind: capture.kind,
                severity: capture.severity,
                triggeredAt: capture.triggeredAt,
                thresholdValue: capture.thresholdValue,
                triggerValue: capture.triggerValue,
                triggerUnit: capture.triggerUnit,
                conditionDuration: capture.conditionDuration,
                samples: capture.samples
            ))
        }
        if !completed.isEmpty {
            try? context.save()
            try? HistoryLocalStore.enforceRetention(in: context)
            refreshPendingCount()
            completed.forEach { onIncidentStored?($0.samples.count) }
        }
    }

    private func freshLocation(at timestamp: Date) -> RecordedLocation? {
        guard locationTracker.isEnabled,
              let location = locationTracker.latestLocation,
              abs(timestamp.timeIntervalSince(location.timestamp)) <= 30 else { return nil }
        return location
    }

    private func refreshPendingCount() {
        let descriptor = FetchDescriptor<DriveIncident>(predicate: #Predicate { $0.syncStateRaw != "synced" })
        pendingIncidentCount = (try? context.fetchCount(descriptor)) ?? 0
    }
}

private extension VehicleAlertRules {
    var incidentConfiguration: IncidentCaptureEngine.Configuration {
        var configuration = IncidentCaptureEngine.Configuration.standard
        configuration.cooldown = TimeInterval(cooldownSeconds)
        configuration.lowOilPressureEnabled = lowOilPressureEnabled
        configuration.lowOilPressureBar = minimumOilPressureBar
        configuration.lowOilMinimumRPM = lowOilMinimumRPM
        configuration.lowOilDuration = lowOilDurationSeconds
        configuration.leanUnderBoostEnabled = leanUnderBoostEnabled
        configuration.leanAFR = maximumAFR
        configuration.leanMinimumBoostBar = leanMinimumBoostBar
        configuration.leanDuration = leanDurationSeconds
        configuration.overboostEnabled = overboostEnabled
        configuration.overboostBar = maximumBoostBar
        configuration.overboostDuration = overboostDurationSeconds
        configuration.highCoolantTemperatureEnabled = highCoolantTemperatureEnabled
        configuration.coolantTemperatureCelsius = maximumCoolantCelsius
        configuration.coolantDuration = coolantDurationSeconds
        configuration.highOilTemperatureEnabled = highOilTemperatureEnabled
        configuration.oilTemperatureCelsius = maximumOilTemperatureCelsius
        configuration.oilTemperatureDuration = oilTemperatureDurationSeconds
        configuration.lowBatteryVoltageEnabled = lowBatteryVoltageEnabled
        configuration.lowBatteryVoltage = minimumBatteryVoltage
        configuration.lowBatteryMinimumRPM = lowBatteryMinimumRPM
        configuration.lowBatteryDuration = lowBatteryDurationSeconds
        configuration.lowFuelPressureEnabled = lowFuelPressureEnabled
        configuration.lowFuelPressureBar = minimumFuelPressureBar
        configuration.lowFuelPressureMinimumRPM = lowFuelPressureMinimumRPM
        configuration.lowFuelPressureDuration = lowFuelPressureDurationSeconds
        return configuration
    }
}

File: android/app/src/main/java/it/letscode/tougedash/history/IncidentCaptureEngine.kt
package it.letscode.tougedash.history

import it.letscode.tougedash.data.local.IncidentEntity
import it.letscode.tougedash.data.local.SyncState
import it.letscode.tougedash.data.local.TougeDashDao
import it.letscode.tougedash.model.ActiveAlert
import it.letscode.tougedash.model.IncidentKind
import it.letscode.tougedash.model.IncidentSeverity
import it.letscode.tougedash.model.RecordedLocation
import it.letscode.tougedash.model.TelemetrySnapshot
import it.letscode.tougedash.model.VehicleAlertRules
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.UUID

@Serializable
data class CapturedTelemetryPoint(
    val id: String = UUID.randomUUID().toString(),
    val recordedAt: Long,
    val rpm: Double,
    val boostBar: Double,
    val mapKpa: Double,
    val throttlePercent: Double,
    val coolantCelsius: Double,
    val intakeCelsius: Double,
    val egt1Celsius: Double = 0.0,
    val egt2Celsius: Double = 0.0,
    val oilTemperatureCelsius: Double,
    val oilPressureBar: Double,
    val fuelPressureBar: Double,
    val afr: Double,
    val lambda: Double,
    val batteryVoltage: Double,
    val ignitionDegrees: Double,
    val injectorDutyPercent: Double,
    val speedKph: Double,
    val checkEngineMask: Int,
    val latitude: Double?,
    val longitude: Double?,
    val horizontalAccuracy: Double?,
    val altitude: Double?
)

class IncidentCaptureEngine(
    private val dao: TougeDashDao,
    private val json: Json,
    private val alertSink: (ActiveAlert) -> Unit,
    private val preSeconds: Int = 30,
    private val postSeconds: Int = 60,
    private val sampleRate: Int = 25
) {
    private data class Match(val kind: IncidentKind, val severity: IncidentSeverity, val value: Double, val threshold: Double, val unit: String, val duration: Double)
    private data class Capture(
        val match: Match,
        val triggeredAt: Long,
        val conditionStartedAt: Long,
        var conditionEndedAt: Long?,
        val points: MutableList<CapturedTelemetryPoint>
    )
    private val ring = ArrayDeque<CapturedTelemetryPoint>()
    private val candidates = mutableMapOf<IncidentKind, Long>()
    private val lastAlerts = mutableMapOf<IncidentKind, Long>()
    private val captures = mutableMapOf<IncidentKind, Capture>()
    var rules = VehicleAlertRules()

    suspend fun record(vehicleId: String, sessionId: String?, snapshot: TelemetrySnapshot, location: RecordedLocation? = null, now: Long = System.currentTimeMillis()) {
        if (sessionId == null) return
        val point = snapshot.toCaptured(now, location)
        ring += point
        while (ring.size > preSeconds * sampleRate) ring.removeFirst()
        captures.values.forEach {
            it.points += point
            if (it.conditionEndedAt == null && !isActive(it.match.kind, snapshot)) it.conditionEndedAt = now
        }

        matches(snapshot).forEach { match ->
            val activeSince = if (isActive(match.kind, snapshot)) candidates.getOrPut(match.kind) { now } else null
            if (activeSince == null) candidates.remove(match.kind)
            val cooldownReady = now - (lastAlerts[match.kind] ?: 0L) >= rules.cooldownSeconds * 1_000L
            if (activeSince != null && now - activeSince >= (match.duration * 1_000).toLong() && cooldownReady && match.kind !in captures) {
                val capture = Capture(match, now, activeSince, null, ring.toMutableList())
                captures[match.kind] = capture
                lastAlerts[match.kind] = now
                alertSink(ActiveAlert(match.kind, match.severity, now, match.value, match.threshold, match.unit))
            }
        }

        val completed = captures.filterValues { now - it.triggeredAt >= postSeconds * 1_000L }.toMap()
        completed.forEach { (kind, capture) ->
            persist(vehicleId, sessionId, capture)
            captures.remove(kind)
        }
    }

    suspend fun finish(vehicleId: String, sessionId: String?) {
        if (sessionId != null) captures.values.toList().forEach { persist(vehicleId, sessionId, it) }
        captures.clear(); candidates.clear(); ring.clear()
    }

    private suspend fun persist(vehicleId: String, sessionId: String, capture: Capture) {
        val trigger = capture.points.minByOrNull { kotlin.math.abs(it.recordedAt - capture.triggeredAt) }
        val first = capture.points.firstOrNull()?.recordedAt ?: capture.triggeredAt
        val last = capture.points.lastOrNull()?.recordedAt ?: capture.triggeredAt
        dao.upsertIncident(
            IncidentEntity(
                id = UUID.randomUUID().toString(), vehicleHardwareId = vehicleId, sessionId = sessionId,
                kind = capture.match.kind.name, severity = capture.match.severity.name,
                triggeredAt = capture.triggeredAt, captureStartedAt = first, captureEndedAt = last,
                sampleCount = capture.points.size,
                sampleRateHz = if (last > first) (capture.points.size - 1) * 1000.0 / (last - first) else sampleRate.toDouble(),
                triggerValue = capture.match.value, thresholdValue = capture.match.threshold,
                triggerUnit = capture.match.unit, triggerRpm = trigger?.rpm ?: 0.0,
                triggerBoostBar = trigger?.boostBar ?: 0.0, triggerAfr = trigger?.afr ?: 0.0,
                triggerSpeedKph = trigger?.speedKph ?: 0.0,
                triggerFuelPressureBar = trigger?.fuelPressureBar ?: 0.0,
                conditionDurationMillis = ((capture.conditionEndedAt ?: last) - capture.conditionStartedAt).coerceAtLeast(0),
                latitude = trigger?.latitude,
                longitude = trigger?.longitude, encodedSamples = json.encodeToString(capture.points),
                syncState = SyncState.PENDING_UPLOAD
            )
        )
    }

    private fun matches(s: TelemetrySnapshot) = listOfNotNull(
        Match(IncidentKind.LOW_OIL_PRESSURE, IncidentSeverity.CRITICAL, s.oilPressureBar, rules.minimumOilPressureBar, "bar", rules.lowOilDurationSeconds).takeIf { rules.lowOilPressureEnabled },
        Match(IncidentKind.LEAN_UNDER_BOOST, IncidentSeverity.WARNING, s.afr, rules.maximumAfr, "AFR", rules.leanDurationSeconds).takeIf { rules.leanUnderBoostEnabled },
        Match(IncidentKind.OVERBOOST, IncidentSeverity.WARNING, s.boostBar, rules.maximumBoostBar, "bar", rules.overboostDurationSeconds).takeIf { rules.overboostEnabled },
        Match(IncidentKind.HIGH_COOLANT_TEMPERATURE, IncidentSeverity.CRITICAL, s.coolantCelsius, rules.maximumCoolantCelsius, "°C", rules.coolantDurationSeconds).takeIf { rules.highCoolantTemperatureEnabled },
        Match(IncidentKind.HIGH_OIL_TEMPERATURE, IncidentSeverity.CRITICAL, s.oilTemperatureCelsius, rules.maximumOilTemperatureCelsius, "°C", rules.oilTemperatureDurationSeconds).takeIf { rules.highOilTemperatureEnabled },
        Match(IncidentKind.LOW_FUEL_PRESSURE, IncidentSeverity.WARNING, s.fuelPressureBar, rules.minimumFuelPressureBar, "bar", rules.lowFuelPressureDurationSeconds).takeIf { rules.lowFuelPressureEnabled },
        Match(IncidentKind.LOW_BATTERY_VOLTAGE, IncidentSeverity.WARNING, s.batteryVoltage, rules.minimumBatteryVoltage, "V", rules.lowBatteryDurationSeconds).takeIf { rules.lowBatteryVoltageEnabled },
        Match(IncidentKind.CHECK_ENGINE, IncidentSeverity.CRITICAL, s.checkEngineMask.toDouble(), 0.0, "CEL", 0.1).takeIf { s.hasCheckEngine }
    )

    private fun isActive(kind: IncidentKind, s: TelemetrySnapshot): Boolean = when (kind) {
        IncidentKind.LOW_OIL_PRESSURE -> s.rpm >= rules.lowOilMinimumRpm && s.oilPressureBar > 0 && s.oilPressureBar < rules.minimumOilPressureBar
        IncidentKind.LEAN_UNDER_BOOST -> s.boostBar >= rules.leanMinimumBoostBar && s.afr > rules.maximumAfr
        IncidentKind.OVERBOOST -> s.boostBar > rules.maximumBoostBar
        IncidentKind.HIGH_COOLANT_TEMPERATURE -> s.coolantCelsius >= rules.maximumCoolantCelsius
        IncidentKind.HIGH_OIL_TEMPERATURE -> s.oilTemperatureCelsius >= rules.maximumOilTemperatureCelsius
        IncidentKind.LOW_FUEL_PRESSURE -> s.rpm >= rules.lowFuelPressureMinimumRpm && s.fuelPressureBar > 0 && s.fuelPressureBar < rules.minimumFuelPressureBar
        IncidentKind.LOW_BATTERY_VOLTAGE -> s.rpm >= rules.lowBatteryMinimumRpm && s.batteryVoltage > 0 && s.batteryVoltage < rules.minimumBatteryVoltage
        IncidentKind.CHECK_ENGINE -> s.hasCheckEngine
    }

    private fun TelemetrySnapshot.toCaptured(now: Long, location: RecordedLocation?) = CapturedTelemetryPoint(
        recordedAt = now, rpm = rpm, boostBar = boostBar, mapKpa = mapKpa,
        throttlePercent = throttlePercent, coolantCelsius = coolantCelsius, intakeCelsius = intakeCelsius,
        egt1Celsius = egt1Celsius, egt2Celsius = egt2Celsius,
        oilTemperatureCelsius = oilTemperatureCelsius, oilPressureBar = oilPressureBar,
        fuelPressureBar = fuelPressureBar, afr = afr, lambda = lambda, batteryVoltage = batteryVoltage,
        ignitionDegrees = ignitionDegrees, injectorDutyPercent = injectorDutyPercent, speedKph = speedKph,
        checkEngineMask = checkEngineMask, latitude = location?.latitude, longitude = location?.longitude,
        horizontalAccuracy = location?.horizontalAccuracy, altitude = location?.altitude
    )
}
